File: app/src/main/java/ca/judacribz/gainzassist/activities/how_to_videos/HowToVideos.java
package ca.judacribz.gainzassist.activities.how_to_videos;


import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import butterknife.BindView;
import com.google.android.youtube.player.YouTubeInitializationResult;
import com.google.android.youtube.player.YouTubePlayer;

import java.util.ArrayList;

import ca.judacribz.gainzassist.R;
import com.google.android.youtube.player.YouTubePlayerSupportFragment;
import com.miguelcatalan.materialsearchview.MaterialSearchView;

import static ca.judacribz.gainzassist.activities.start_workout.StartWorkout.EXTRA_HOW_TO_VID;
import static ca.judacribz.gainzassist.util.UI.setInitView;
import static ca.judacribz.gainzassist.util.UI.setToolbar;

public class HowToVideos extends AppCompatActivity implements
        MaterialSearchView.OnQueryTextListener,
        SearchVideosTask.YouTubeSearchObserver,
       ThumbnailAdapter.VideoClickObserver,
       YouTubePlayer.OnInitializedListener,
       YouTubePlayer.PlayerStateChangeListener {

    // Constants
    // --------------------------------------------------------------------------------------------
    public static final String EXTRA_VIDEO_ID =
            "ca.judacribz.gainzassist.act_how_to_videos.EXTRA_VIDEO_ID";
    public static final String SEARCH_SPACE_STR = "%20";

    String URL = "https://www.googleapis.com/youtube/v3/search?" +  // default youtube search url
                 "part=snippet&" +                                  // search resource
                 "fields=items(id/videoId,snippet/title)&" +        // needed fields
                 "maxResults=10&" +                                 // number of results to show
                 "q=";                                              // search text
    // --------------------------------------------------------------------------------------------

    // Global Vars
    // --------------------------------------------------------------------------------------------
    RecyclerView rvVideoList;
    ThumbnailAdapter thumbnailAdapter;
    LinearLayoutManager rvLayoutManager;
    SearchVideosTask task;

    YouTubePlayer player;
    YouTubePlayerSupportFragment ytpFmt;
    FragmentManager fmtMgr;
    FragmentTransaction fmtTxn;
    String videoId, exerciseName;

    boolean isFullScreen = false;
    boolean backPressed = false;

    @BindView(R.id.msvHowToVids) MaterialSearchView searchView;
    //---------------------------------------------------------------------------------------------

    // AppCompatActivity Override
    ///////////////////////////////////////////////////////////////////////////////////////////////
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        exerciseName = getIntent().getStringExtra(EXTRA_HOW_TO_VID);
        setInitView(this, R.layout.activity_how_to_videos, "How To " + exerciseName, true);

        fmtMgr = getSupportFragmentManager();

        rvVideoList = (RecyclerView) findViewById(R.id.rv_video_list);
        ytpFmt = (YouTubePlayerSupportFragment) getSupportFragmentManager().findFragmentById(R.id.fmt_youtube);

        handleFmt(false);

        executeSearch("how to " + exerciseName);

        rvLayoutManager = new LinearLayoutManager(this);
        rvVideoList.setLayoutManager(rvLayoutManager);
        rvVideoList.setHasFixedSize(true);
    }

    @Override
    protected void onResume() {
        super.onResume();

        searchView.setOnQueryTextListener(this);
        searchView.setVoiceSearch(true);
    }

    @Override
    public boolean onSupportNavigateUp() {

        if (ytpFmt.isHidden()) {
            finish();
        } else {
            handleFmt(false);
        }

        return true;
    }

    @Override
    public void onBackPressed() {
        if (!ytpFmt.isHidden()) {
            handleFmt(false);
        } else {
            super.onBackPressed();
        }
    }

    /* Handles showing or hiding the youtube player fragment */
    void handleFmt(boolean showFmt) {
        fmtTxn = fmtMgr.beginTransaction();

        if (showFmt) {
            ytpFmt.initialize(getString(R.string.google_api_key), this);
            fmtTxn.show(ytpFmt);

        } else {
            if (player != null) {
                if (isFullScreen) {
                    backPressed = true;
                    player.setFullscreen(false);
                } else {
                    player.release();
                }
            }
            fmtTxn.hide(ytpFmt);
        }

        fmtTxn.commitNow();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu mainMenu) {
        getMenuInflater().inflate(R.menu.menu_how_to, mainMenu);

        searchView.setMenuItem(mainMenu.findItem(R.id.act_search));

        return super.onCreateOptionsMenu(mainMenu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {

        switch (item.getItemId()) {
            case R.id.act_search:
                break;
        }
        return super.onOptionsItemSelected(item);
    }
    //AppCompatActivity//Override//////////////////////////////////////////////////////////////////


    // MaterialSearchView.OnQueryTextListener Override
    ///////////////////////////////////////////////////////////////////////////////////////////////
    @Override
    public boolean onQueryTextChange(String query) {
        return false;
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
        executeSearch(query);

        return false;
    }
    //MaterialSearchView.OnQueryTextListener//Override/////////////////////////////////////////////


    private void executeSearch(String query) {
        if (task != null) {
            task.setYouTubeSearchObserver(null);
        }
        task = new SearchVideosTask();
        task.setYouTubeSearchObserver(this);
        task.execute(URL.concat(query.replaceAll("\\s+", SEARCH_SPACE_STR))
                .concat("&key=")
                .concat(getString(R.string.google_api_key)));
    }


    // Interface Callback: SearchVideosTask.YouTubeSearchObserver Override
    ///////////////////////////////////////////////////////////////////////////////////////////////
    @Override
    public void videoSearchDataReceived(ArrayList<String> videoIds, ArrayList<String> videoTitles) {
        if (videoIds.size() > 0) {
            thumbnailAdapter = new ThumbnailAdapter(videoIds, videoTitles);
            rvVideoList.setAdapter(thumbnailAdapter);
            thumbnailAdapter.setVideoClickObserver(this);

        } else {
            Snackbar.make(rvVideoList, "No video results", Snackbar.LENGTH_SHORT).show();
        }
    }

    //SearchVideosTask.YouTubeSearchObserver//Override/////////////////////////////////////////////


    // Interface Callback: ThumbnailAdapter.VideoClickObserver Override
    ///////////////////////////////////////////////////////////////////////////////////////////////
    @Override
    public void onVideoClick(final String videoId) {
        this.videoId = videoId;
        handleFmt(true);
    }
    //ThumbnailAdapter.VideoClickObserver//Override////////////////////////////////////////////////

    // Interface Callback: YouTubePlayer.OnInitializedListener Override
    ///////////////////////////////////////////////////////////////////////////////////////////////
    @Override
    public void onInitializationSuccess(YouTubePlayer.Provider provider,
                                        final YouTubePlayer player,
                                        boolean b) {
        this.player = player;
        player.setPlayerStateChangeListener(this);
        player.setOnFullscreenListener(new YouTubePlayer.OnFullscreenListener() {
            @Override
            public void onFullscreen(boolean isfullScreen) {

                HowToVideos.this.isFullScreen = isfullScreen;
                if (!isfullScreen) {
                    if (backPressed) {
                        player.release();
                    }
                } else {
                    Toast.makeText(HowToVideos.this, "Fullscreen", Toast.LENGTH_SHORT).show();

                    ytpFmt.initialize(getString(R.string.google_api_key), HowToVideos.this);
                }
            }

        });
        player.setShowFullscreenButton(true);
        player.loadVideo(videoId);
    }

    @Override
    public void onInitializationFailure(YouTubePlayer.Provider provider,
                                        YouTubeInitializationResult initializationResult) {
    }
    //YouTubePlayer.OnInitializedListener//Override////////////////////////////////////////////////


    // YouTubePlayer.PlayerStateChangeListener Override
    ///////////////////////////////////////////////////////////////////////////////////////////////
    @Override
    public void onLoading() {
    }

    @Override
    public void onLoaded(String s) {

    }

    @Override
    public void onAdStarted() {
    }

    @Override
    public void onVideoStarted() {
    }

    @Override
    public void onVideoEnded() {
        handleFmt(false);
    }

    @Override
    public void onError(YouTubePlayer.ErrorReason errorReason) {
    }
    //YouTubePlayer.PlayerStateChangeListener//Override////////////////////////////////////////////
}

File: app/src/main/java/ca/judacribz/gainzassist/activities/how_to_videos/HowToVideos.kt
package ca.judacribz.gainzassist.activities.how_to_videos

import android.content.Intent
import android.net.Uri
import android.content.pm.PackageManager
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import ca.judacribz.gainzassist.BuildConfig
import ca.judacribz.gainzassist.R
import ca.judacribz.gainzassist.activities.start_workout.StartWorkout.Companion.EXTRA_HOW_TO_VID
import ca.judacribz.gainzassist.databinding.ActivityHowToVideosBinding
import ca.judacribz.gainzassist.util.UI.setInitTheme
import ca.judacribz.gainzassist.util.UI.setToolbar
import com.google.android.material.snackbar.Snackbar
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.PlayerConstants
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.YouTubePlayer
import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.listeners.AbstractYouTubePlayerListener
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONException
import org.json.JSONObject
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import java.util.ArrayList
import java.util.HashMap

class HowToVideos : AppCompatActivity(),
    SearchView.OnQueryTextListener,
    ThumbnailAdapter.VideoClickObserver {

    companion object {
        const val EXTRA_VIDEO_ID = "ca.judacribz.gainzassist.act_how_to_videos.EXTRA_VIDEO_ID"
    }

    private var thumbnailAdapter: ThumbnailAdapter? = null

    private var youTubePlayer: YouTubePlayer? = null
    private var pendingVideoId: String? = null
    private var videoId: String? = null
    private var exerciseName: String? = null

    private val queryCache = HashMap<String, Pair<ArrayList<String>, ArrayList<String>>>()
    private var activeQuery: String? = null

    private lateinit var binding: ActivityHowToVideosBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        exerciseName = intent.getStringExtra(EXTRA_HOW_TO_VID)
        setInitTheme(this)
        binding = ActivityHowToVideosBinding.inflate(layoutInflater)
        setContentView(binding.root)
        setToolbar(this, "How To ${exerciseName ?: ""}", true)

        lifecycle.addObserver(binding.ypvPlayer)

        if (exerciseName.isNullOrBlank()) {
            Snackbar.make(binding.rvVideoList, "Search for an exercise video.", Snackbar.LENGTH_LONG).show()
        } else {
            executeSearch("how to $exerciseName")
        }

        binding.rvVideoList.layoutManager = LinearLayoutManager(this)
        binding.rvVideoList.setHasFixedSize(true)

        binding.ypvPlayer.addYouTubePlayerListener(object : AbstractYouTubePlayerListener() {
            override fun onReady(player: YouTubePlayer) {
                youTubePlayer = player
                pendingVideoId?.let {
                    player.loadVideo(it, 0f)
                    pendingVideoId = null
                }
            }

            override fun onError(player: YouTubePlayer, error: PlayerConstants.PlayerError) {
                super.onError(player, error)
                val currentVideoId = videoId
                if (currentVideoId != null) {
                    Snackbar.make(binding.rvVideoList, "Unable to play this video", Snackbar.LENGTH_LONG)
                        .setAction("Open in YouTube") {
                            val intent = Intent(Intent.ACTION_VIEW, Uri.parse("https://www.youtube.com/watch?v=$currentVideoId"))
                            startActivity(intent)
                        }
                        .show()
                } else {
                    Snackbar.make(binding.rvVideoList, "Unable to play this video", Snackbar.LENGTH_SHORT).show()
                }
            }
        })
    }

    override fun onSupportNavigateUp(): Boolean {
        if (binding.ypvPlayer.visibility == View.INVISIBLE || binding.ypvPlayer.visibility == View.GONE) {
            finish()
        } else {
            closePlayer()
        }
        return true
    }

    override fun onBackPressed() {
        if (binding.ypvPlayer.visibility == View.VISIBLE) {
            closePlayer()
        } else {
            super.onBackPressed()
        }
    }

    override fun onCreateOptionsMenu(mainMenu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_how_to, mainMenu)
        val searchItem = mainMenu.findItem(R.id.act_search)
        val searchView = searchItem.actionView as? SearchView
        searchView?.setOnQueryTextListener(this)
        return super.onCreateOptionsMenu(mainMenu)
    }

    override fun onQueryTextChange(query: String): Boolean {
        return false
    }

    override fun onQueryTextSubmit(query: String): Boolean {
        executeSearch(query)
        return false
    }

    override fun onVideoClick(videoId: String) {
        this.videoId = videoId
        binding.ypvPlayer.visibility = View.VISIBLE
        if (youTubePlayer != null) {
            youTubePlayer?.loadVideo(videoId, 0f)
        } else {
            pendingVideoId = videoId
        }
    }

    private fun closePlayer() {
        youTubePlayer?.pause()
        binding.ypvPlayer.visibility = View.INVISIBLE
    }

    private fun executeSearch(query: String) {
        val trimmedQuery = query.trim()

        if (trimmedQuery == activeQuery) {
            return
        }

        if (queryCache.containsKey(trimmedQuery)) {
            val cached = queryCache[trimmedQuery]
            if (cached != null) {
                videoSearchDataReceived(cached.first, cached.second)
            }
            return
        }

        if (BuildConfig.GOOGLE_API_KEY.isBlank()) {
            Snackbar.make(binding.rvVideoList, "Missing Google API Key for video search", Snackbar.LENGTH_LONG).show()
            return
        }

        activeQuery = trimmedQuery

        val uri = Uri.Builder()
            .scheme("https")
            .authority("www.googleapis.com")
            .appendPath("youtube")
            .appendPath("v3")
            .appendPath("search")
            .appendQueryParameter("part", "snippet")
            .appendQueryParameter("fields", "items(id/videoId,snippet/title)")
            .appendQueryParameter("maxResults", "10")
            .appendQueryParameter("q", trimmedQuery)
            .appendQueryParameter("type", "video")
            .appendQueryParameter("videoEmbeddable", "true")
            .appendQueryParameter("safeSearch", "moderate")
            .appendQueryParameter("key", BuildConfig.GOOGLE_API_KEY)
            .build()

        searchYouTube(uri.toString())
    }

    private fun getAppSha1(): String {
        try {
            val packageInfo = packageManager.getPackageInfo(packageName, PackageManager.GET_SIGNATURES)
            val signatures = packageInfo.signatures
            if (signatures != null) {
                for (signature in signatures) {
                    val md = MessageDigest.getInstance("SHA-1")
                    md.update(signature.toByteArray())
                    val digest = md.digest()
                    val hexString = StringBuilder()
                    for (b in digest) {
                        hexString.append(String.format("%02X", b))
                    }
                    return hexString.toString()
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return ""
    }

    private fun searchYouTube(urlString: String) {
        lifecycleScope.launch(Dispatchers.IO) {
            var connection: HttpURLConnection? = null
            var reader: BufferedReader? = null
            var errorMessage: String? = null
            var videoIds: ArrayList<String>? = null
            var videoTitles: ArrayList<String>? = null

            try {
                val url = URL(urlString)
                connection = url.openConnection() as HttpURLConnection
                connection.setRequestProperty("X-Android-Package", packageName)
                connection.setRequestProperty("X-Android-Cert", getAppSha1())
                connection.connect()

                val responseCode = connection.responseCode
                val stream = if (responseCode in 200..299) {
                    connection.inputStream
                } else {
                    connection.errorStream
                }

                if (stream != null) {
                    reader = BufferedReader(InputStreamReader(stream))
                    val buffer = StringBuilder()
                    var line: String?
                    while (reader.readLine().also { line = it } != null) {
                        buffer.append(line).append("\n")
                    }

                    if (responseCode !in 200..299) {
                        val errorBody = buffer.toString()
                        val sanitizedUrl = urlString.replace(Regex("&key=[^&]*"), "&key=***")
                        Log.e("HowToVideos", "Error $responseCode for $sanitizedUrl\nBody: $errorBody")

                        errorMessage = if (errorBody.contains("quotaExceeded") || errorBody.contains("dailyLimitExceeded")) {
                            "YouTube video search quota exceeded. Please try again later."
                        } else {
                            "Unable to load YouTube videos. Check API key or network."
                        }
                    } else {
                        val jsonObject = JSONObject(buffer.toString())
                        val items = jsonObject.optJSONArray("items")

                        videoIds = ArrayList()
                        videoTitles = ArrayList()

                        if (items != null) {
                            for (i in 0 until items.length()) {
                                val item = items.optJSONObject(i) ?: continue
                                val id = item.optJSONObject("id") ?: continue
                                val snippet = item.optJSONObject("snippet") ?: continue

                                val videoId = id.optString("videoId")
                                val title = snippet.optString("title")

                                if (videoId.isNotEmpty()) {
                                    videoIds.add(videoId)
                                    videoTitles.add(title)
                                }
                            }
                        }
                    }
                } else {
                    errorMessage = "Unable to load YouTube videos. Check API key or network."
                }
            } catch (e: IOException) {
                e.printStackTrace()
                errorMessage = "Unable to load YouTube videos. Check API key or network."
            } catch (e: JSONException) {
                e.printStackTrace()
                errorMessage = "Error parsing response"
            } finally {
                connection?.disconnect()
                try {
                    reader?.close()
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }

            withContext(Dispatchers.Main) {
                if (errorMessage != null) {
                    videoSearchFailed(errorMessage)
                } else if (videoIds != null && videoTitles != null) {
                    videoSearchDataReceived(videoIds, videoTitles)
                }
            }
        }
    }

    private fun videoSearchDataReceived(videoIds: ArrayList<String>, videoTitles: ArrayList<String>) {
        val currentQuery = activeQuery
        if (currentQuery != null) {
            queryCache[currentQuery] = Pair(videoIds, videoTitles)
        }
        activeQuery = null

        if (videoIds.size > 0) {
            thumbnailAdapter = ThumbnailAdapter(videoIds, videoTitles)
            binding.rvVideoList.adapter = thumbnailAdapter
            thumbnailAdapter?.setVideoClickObserver(this)
        } else {
            Snackbar.make(binding.rvVideoList, "No video results", Snackbar.LENGTH_SHORT).show()
        }
    }

    private fun videoSearchFailed(message: String) {
        activeQuery = null
        Snackbar.make(binding.rvVideoList, message, Snackbar.LENGTH_LONG).show()
    }
}
